README.rst
pytest-responsemock
===================
https://github.com/idlesign/pytest-responsemock

|release| |lic| |coverage|

.. |release| image:: https://img.shields.io/pypi/v/pytest-responsemock.svg
    :target: https://pypi.python.org/pypi/pytest-responsemock

.. |lic| image:: https://img.shields.io/pypi/l/pytest-responsemock.svg
    :target: https://pypi.python.org/pypi/pytest-responsemock

.. |coverage| image:: https://img.shields.io/coveralls/idlesign/pytest-responsemock/master.svg
    :target: https://coveralls.io/r/idlesign/pytest-responsemock


Description
-----------

*Simplified requests calls mocking for pytest*

Provides ``response_mock`` fixture, exposing simple context manager.

Any request under that manager will be intercepted and mocked according
to one or more ``rules`` passed to the manager. If actual request won't fall
under any of given rules then an exception is raised (by default).

Rules are simple strings, of the pattern: ``HTTP_METHOD URL -> STATUS_CODE :BODY``.


Requirements
------------

* Python 3.6+


Usage
-----

When this package is installed ``response_mock`` is available for ``pytest`` test functions.

.. code-block:: python

    def for_test():
        return requests.get('http://some.domain')


    def test_me(response_mock):

        # Pass response rule as a string,
        # or many rules (to mock consequent requests) as a list of strings/bytes.
        # Use optional `bypass` argument to disable mock conditionally.

        with response_mock('GET http://some.domain -> 200 :Nice', bypass=False):

            result = for_test()

            assert result.ok
            assert result.content == b'Nice'
            
        # mock consequent requests
        with response_mock([
            'GET http://some.domain -> 200 :Nice',
            'GET http://other.domain -> 200 :Sweet',
        ]):
            for_test()
            requests.get('http://other.domain')


Use with ``pytest-datafixtures``:

.. code-block:: python

    def test_me(response_mock, datafix_read):

        with response_mock(f"GET http://some.domain -> 200 :{datafix_read('myresponse.html')}"):
            ...


Describe response header fields using multiline strings:

.. code-block:: python

    with response_mock(
        '''
        GET http://some.domain

        Allow: GET, HEAD
        Content-Language: ru

        -> 200 :OK
        '''
    ):
        ...

Test json response:

.. code-block:: python

    response = json.dumps({'key': 'value', 'another': 'yes'})

    with response_mock(f'POST http://some.domain -> 400 :{response}'):
        ...

To test binary response pass rule as bytes:

.. code-block:: python

    with response_mock(b'GET http://some.domain -> 200 :' + my_bytes):
        ...

Access underlying RequestsMock (from ``responses`` package) as ``mock``:

.. code-block:: python

    with response_mock('HEAD http://some.domain -> 200 :Nope') as mock:

        mock.add_passthru('http://other.domain')



CHANGELOG
pytest-responsemock changelog
=============================


v1.1.1 [2022-03-10]
-------------------
* Made responses 0.19.0 compatible.


v1.1.0 [2022-02-04]
-------------------
! Dropped support for Python 3.6.


v1.0.1 [2020-10-10]
-------------------
* Handle setting 'Content-Type' header correctly.


v1.0.0 [2020-07-26]
-------------------
+ Add support for binary responses.


v0.2.0 [2020-03-31]
-------------------
+ Added support for response header fileds.


v0.1.1 [2020-03-10]
-------------------
* Plugin made friendly to shared virtual envs.


v0.1.0 [2020-03-05]
-------------------
+ Basic functionality.

tests/test_module.py
import requests


def run_get():
    response = requests.get('http://yandex.ru', allow_redirects=False)
    return response


def test_oneline(response_mock):

    with response_mock('GET http://yandex.ru -> 200 :Nice'):
        result = run_get()
        assert result.ok
        assert result.content == b'Nice'


def test_binary(response_mock):

    with response_mock(
            b'GET http://yandex.ru \n'
            b'Allow: GET, HEAD\n'
            b'Content-Language: ru\n'
            b'-> 200 :\xd1\x82\xd0\xb5\xd1\x81\xd1\x82'
    ):
        result = requests.get('http://yandex.ru', allow_redirects=False)
        assert result.ok
        assert result.content.decode() == 'тест'
        assert result.headers['Content-Language'] == 'ru'


def test_header_fields(response_mock):

    with response_mock('''
        GET http://yandex.ru
        
        Content-Type: image/png
        Cache-Control: no-cache,no-store,max-age=0,must-revalidate
        Set-Cookie: key1=val1
        
        -> 200 :Nicer
    '''):
        result = run_get()
        assert result.ok
        assert result.content == b'Nicer'
        assert dict(result.headers) == {
            'Content-Type': 'image/png',
            'Cache-Control': 'no-cache,no-store,max-age=0,must-revalidate',
            'Set-Cookie': 'key1=val1'
        }
        assert dict(result.cookies) == {'key1': 'val1'}


def test_many_lines(response_mock):

    with response_mock([
        'GET http://yandex.ru -> 200 :Nice',
        '',  # Support empty rule for debug actual request in tests.
    ]):
        result = run_get()
        assert result.content == b'Nice'


def test_status(response_mock):

    with response_mock('GET http://yandex.ru -> 500 :Bad'):
        result = run_get()
        assert not result.ok
        assert result.content == b'Bad'


def test_bypass(response_mock):

    with response_mock('GET http://yandex.ru -> 500 :Nice', bypass=True):
        result = run_get()
        assert result.status_code == 301  # https redirect


responsemock/entry.py
import pytest

from .utils import response_mock as response_mock_


@pytest.fixture
def response_mock() -> response_mock_:
    """Fixture exposing simple context manager to mock responses for `requests` package.

    Any request under that manager will be intercepted and mocked according
    to one or more ``rules`` passed to the manager. If actual request won't fall
    under any of given rules then an exception is raised (by default).

    Rules are simple strings, of pattern: ``HTTP_METHOD URL -> STATUS_CODE :BODY``.

    Example::

        def test_me(response_mock):


            json_response = json.dumps({'key': 'value', 'another': 'yes'})

            with response_mock([

                'GET http://a.b -> 200 :Nice',

                f'POST http://some.domain -> 400 :{json_response}'

                '''
                GET https://some.domain

                Allow: GET, HEAD
                Content-Language: ru

                -> 200 :OK
                '''

            ], bypass=False) as mock:

                mock.add_passthru('http://c.d')

                this_mades_requests()


    """
    return response_mock_


responsemock/utils.py
from contextlib import contextmanager
from textwrap import dedent
from typing import Union, List, Generator, Optional

if False:  # pragma: nocover
    from responses import RequestsMock  # noqa


@contextmanager
def response_mock(
        rules: Union[List[str], str, List[bytes], bytes],
        *,
        bypass: bool = False,
        **kwargs
) -> Generator[Optional['RequestsMock'], None, None]:
    """Simple context manager to mock responses for `requests` package.

    Any request under that manager will be intercepted and mocked according
    to one or more ``rules`` passed to the manager. If actual request won't fall
    under any of given rules then an exception is raised (by default).

    Rules are simple strings, of pattern: ``HTTP_METHOD URL -> STATUS_CODE :BODY``.

    Example::

        def test_me(response_mock):


            json_response = json.dumps({'key': 'value', 'another': 'yes'})

            with response_mock([

                'GET http://a.b -> 200 :Nice',

                b'GET http://a.b/binary/ -> 200 :\xd1\x82\xd0\xb5\xd1\x81\xd1\x82',

                f'POST http://some.domain -> 400 :{json_response}'

                '''
                GET https://some.domain

                Allow: GET, HEAD
                Content-Language: ru

                -> 200 :OK
                '''

            ], bypass=False) as mock:

                mock.add_passthru('http://c.d')

                this_makes_requests()

    :param rules: One or several rules for response.
    :param bypass: Whether to to bypass (disable) mocking.
    :param kwargs: Additional keyword arguments to pass to `RequestsMock`.

    """
    from responses import RequestsMock
    try:
        # In responses 0.19.0, UNSET was renamed to _UNSET
        from responses import _UNSET as UNSET
    except ImportError:
        from responses import UNSET

    if bypass:

        yield

    else:

        with RequestsMock(**kwargs) as mock:

            def enc(val):
                return val.encode() if is_binary else val

            def dec(val):
                return val.decode() if is_binary else val

            if isinstance(rules, (str, bytes)):
                rules = [rules]

            for rule in rules:

                if not rule:
                    continue

                is_binary = isinstance(rule, bytes)

                if not is_binary:
                    rule = dedent(rule)

                rule = rule.strip()
                directives, _, response = rule.partition(enc('->'))

                headers = {}

                if (enc('\n')) in directives:
                    directives, *headers_block = directives.splitlines()

                    for header_line in headers_block:
                        header_line = header_line.strip()

                        if not header_line:
                            continue

                        key, _, val = header_line.partition(enc(':'))
                        val = val.strip()

                        if val:
                            headers[dec(key.strip())] = dec(val)

                content_type = headers.pop('Content-Type', UNSET)

                directives = list(
                    filter(
                        None,
                        map(
                            (bytes if is_binary else str).strip,
                            directives.split(enc(' '))
                        )
                    ))

                assert len(directives) == 2, (
                    f'Unsupported directives: {directives}. Expected: HTTP_METHOD URL')

                status, _, response = response.partition(enc(':'))

                status = int(status.strip())

                mock.add(
                    method=dec(directives[0]),
                    url=dec(directives[1]),
                    body=response,
                    status=status,
                    adding_headers=headers or None,
                    content_type=content_type,
                )

            yield mock
